example/common/flat_stream.hpp

```diff
@@ -0,0 +1,567 @@
+        flat_stream<T>& s, TeardownHandler&& handler);
+};
+
+//------------------------------------------------------------------------------
+
+template<class NextLayer>
+template<class ConstBufferSequence, class Handler>
+class flat_stream<NextLayer>::write_op
+    : public boost::asio::coroutine
+{
+    using alloc_type = typename 
+        boost::asio::associated_allocator_t<Handler>::template
+            rebind<char>::other;
+
+    using const_iterator = decltype(std::declval<
+        ConstBufferSequence const&>().begin());
+
+    static std::size_t constexpr limit = 64 * 1024;
+
+    struct data
+    {
+        flat_stream<NextLayer>& s;
+        ConstBufferSequence b;
+        const_iterator first;
+        const_iterator last;
+        std::size_t bytes_transferred = 0;
+        alloc_type alloc;
+        char* p = nullptr;
+        std::size_t n = 0;
+
+        data(
+            Handler const& h,
+            flat_stream<NextLayer>& s_,
+            ConstBufferSequence const& b_)
+            : s(s_)
+            , b(b_)
+            , first(boost::asio::buffer_sequence_begin(b))
+            , last(boost::asio::buffer_sequence_end(b))
+            , alloc((boost::asio::get_associated_allocator)(h))
+        {
+        }
+
+        ~data()
+        {
+            if(p)
+                alloc.deallocate(p, n);
+        }
+    };
+
+    boost::beast::handler_ptr<data, Handler> d_;
+
+public:
+    write_op(write_op&&) = default;
+
+    template<class DeducedHandler>
+    write_op(
+        flat_stream<NextLayer>& s,
+        ConstBufferSequence const& b,
+        DeducedHandler&& h)
+        : d_(std::forward<DeducedHandler>(h), s, b)
+    {
+    }
+
+    using allocator_type =
+        boost::asio::associated_allocator_t<Handler>;
+
+    allocator_type
+    get_allocator() const noexcept
+    {
+        return (boost::asio::get_associated_allocator)(d_.handler());
+    }
+
+    using executor_type = boost::asio::associated_executor_t<
+        Handler, decltype(std::declval<NextLayer&>().get_executor())>;
+
+    executor_type
+    get_executor() const noexcept
+    {
+        return (boost::asio::get_associated_executor)(
+            d_.handler(), d_->s.get_executor());
+    }
+
+    void
+    operator()(
+        boost::system::error_code ec,
+        std::size_t bytes_transferred)
+    {
+        auto& d = *d_;
+        BOOST_ASIO_CORO_REENTER(*this)
+        {
+            do
+            {
+                BOOST_ASIO_CORO_YIELD
+                {
+                    auto const result = coalesce(d.first, d.last, limit);
+                    auto const b = buffers_iterator_range(d.first, result.last);
+                    if(result.copy)
+                    {
+                        if(d.p)
+                        {
+                            d.alloc.deallocate(d.p, d.n);
+                            d.p = nullptr;
+                        }
+                        d.p = d.alloc.allocate(result.size);
```

Comment:
Is there a reason that the op performs an allocation on every iteration, rather than once? if we need to allocate memory, we can just allocate it once, using `limit` bytes.

Reply:
This is a straightforward implementation which doesn't try to get too clever with optimizations. The cost of the allocation is very small compared to the cost of invoking the `ssl::stream` initiating functions. So it isn't clear that allocating `limit` bytes and keeping it around is advantageous.